Paxifax.cs

```diff
@@ -23,96 +23,125 @@
-    // But this is fine, because there might be interface for which we don't want to process
-    // their incoming packets, but we want to be able to forward packets to them nonetheless.
-    public static int no_interfaces;
-
-    // Array "maps" from device offset to the device object.
-    public static ICaptureDevice[] deviceMap;
-    // Map from device name (e.g., "eth1") to device offset.
-    public static Dictionary<string, int> rdeviceMap = new Dictionary<string, int>();
-    // Map from device offset to the name of its handler.
-    public static string[] interface_lead_handler;
-    public static PacketProcessor[] interface_lead_handler_obj;
-
-    // FIXME better to link to function (rather than have indirection) to speed things up at runtime.
-    // The file containing the catalogue of network interfaces.
-    public static string config_filename;
-    // The assembly containing the packet handlers that the user wishess to use.
-    public static string assembly_filename;
-    public static Assembly assembly;
-
-    public static List<NetworkInterfaceConfig> config;
-
-    public static string resolve_config_parameter (int port_no, string key) {
-      NetworkInterfaceConfig port_conf;
-      try {
-        port_conf = config[port_no];
-      } catch (ArgumentOutOfRangeException e) {
-        throw (new Exception ("resolve_config_parameter: port_no > config size, since " + port_no.ToString() + " > " +
-              config.Count.ToString()));
-      }
-
-      if (port_conf.environment == null)
+  internal static class PacketProcessorHelper
+  {
+    public readonly static Type[] AllowedConstructorParameterTypes = new Type[]
+      {
+        typeof(Boolean),  typeof(Byte),   typeof(SByte),  typeof(UInt16),   typeof(Int16),
+        typeof(UInt32),   typeof(Int32),  typeof(UInt64), typeof(Int64),    typeof(Decimal),
+        typeof(Single),   typeof(Double), typeof(String), typeof(DateTime),
+        typeof(TimeSpan), typeof(System.Net.IPAddress), typeof(PhysicalAddress)
+      };
+    public static bool IsAllowedConstructorParameterType(Type ty)
+    {
+      // Allow nullable types
+      if (ty.IsGenericType && ty.GetGenericTypeDefinition() == typeof(Nullable<>))
+        ty = Nullable.GetUnderlyingType(ty);
+      return AllowedConstructorParameterTypes.Contains(ty);
+    }
+    public static object ConvertConstructorParameter(Type ty, string s)
+    {
+      if (ty == typeof(string))
+        return s;
+      else if (ty == typeof(System.Net.IPAddress))
+        return System.Net.IPAddress.Parse(s);
+      else if (ty == typeof(PhysicalAddress))
+        return PhysicalAddress.Parse(s.ToUpper().Replace(':', '-'));
+      else if (ty == typeof(TimeSpan))
+        return TimeSpan.Parse(s);
+      else
+        // Convert to primitives + DateTime
+        return ((IConvertible)s).ToType(ty, System.Globalization.CultureInfo.CurrentCulture); // NOTE throws InvalidCastException
```

Comment:
I wanted this method to be more of a utility method, and preserve the error messages for the caller. Note that the method calls before it can throw FormatException. There is some user feedback in InstantiatePacketProcessor on line 134.